File: Sources/Sqlite2CoreData/SQCDMigrationManager.m
//
//  CDMMigrationManager.m
//  CoreDataMigration
//
//  Created by Tapasya on 20/08/13.
//  Copyright (c) 2013 Tapasya. All rights reserved.
//

#import "SQCDMigrationManager.h"
@import FMDB;
@import SQCDHelper;
#import <CoreData/CoreData.h>
#import "SqliteExtractor.h"



@implementation SQCDMigrationManager

- (instancetype)init
{
    self = [super init];
    if (self) {
        self->_dbPath = nil;
        self->_tableDictionary = nil;
    }
    return self;
}

-(void)dealloc {
    if (self.dataBase) {
        [self.dataBase close];
    }
}

#pragma mark - Migration

+(BOOL) startDataMigrationWithDBPath:(NSString*) dbPath
                            momdPath:(NSString*) momnPath
                          outputPath:(NSString*) outputPath
                              helper:(SQCDDatabaseHelper*) helper
{
    SQCDMigrationManager* manager = SQCDMigrationManager.new;
    manager.helper = helper;
    manager.dbPath = dbPath;

    
    FMDatabase* database = [manager openDatabase];
    manager.tableDictionary = [manager.helper fetchTableInfos:dbPath];
    SQCDCoreDataManager* cdm = [[SQCDCoreDataManager alloc] initWithModelPath:momnPath outputDirectory:outputPath];
    NSManagedObjectContext* moc = [cdm managedObjectContext];
    
    [manager migrateTableDataFromDatabase:database
                               toManagedObjectContext:moc
                                    withRelationships:YES];
    
    return YES;
}

#pragma mark - data and relationship migration

- (BOOL) migrateTableDataFromDatabase:(FMDatabase*) database
               toManagedObjectContext:(NSManagedObjectContext*) moc
                    withRelationships:(BOOL) migrateRelationships
{
    NSDictionary *tablesDict = [self tableDictionary];
    __block __weak NSError* error;
    
    for (SQCDTableInfo* tableInfo in [tablesDict allValues]) {
        NSString* tableName =  tableInfo.sqliteName;
        
        @autoreleasepool {
            if (![tableName isEqualToString:@"sqlite_sequence"] && ![tableInfo isManyToMany]){
                
                NSLog(@"***************  Started migration for table %@  ****************", tableInfo.sqliteName);
                
                FMResultSet *results = [database executeQuery:[NSString stringWithFormat:@"select * from %@", tableName]];
                while([results next]) {
                    @autoreleasepool {
                        NSManagedObject *entity =  [self createEntityFromResultSet:results
                                                                                    withTableInfo:tableInfo
                                                                           inManagedObjectContext:moc
                                                                            shouldCreateManyToMany:YES];
                        [moc save:&error];
                        if ( nil != error){
                            NSLog(@"Error while saving %@, %@", entity.entity.name, [error localizedDescription]);
                            for (NSError* detailedError in [[error userInfo] valueForKey:NSDetailedErrorsKey]) {
                                NSLog(@"Error while saving %@", [detailedError debugDescription]);
                            }
                        }

                    }
                }
                NSLog(@"***************  Ended migration for table %@  ****************", tableInfo.sqliteName);
            }
        }
    }
    
    return error != nil;
}

#pragma mark - NSManagedObject Creation

- (NSArray*) fetchEntity:(NSString*) entityName
     primaryPropertyName:(NSString*) propertyName
    primaryPropertyValue:(int) primaryId
  inManagedObjectContext:(NSManagedObjectContext*) moc
{
    NSArray* toEntities = nil;
    @try {
        NSError* error;
        NSFetchRequest *toFetchRequest = [[NSFetchRequest alloc] init];
        [toFetchRequest setEntity:[NSEntityDescription entityForName:entityName inManagedObjectContext:moc]];
        NSString* predicateString = [NSString stringWithFormat:@"(%@ IN %@)", propertyName, @"%@"];
        [toFetchRequest setPredicate: [NSPredicate predicateWithFormat:predicateString,[NSArray arrayWithObject:[NSNumber numberWithInteger:primaryId]]]];
       toEntities = [moc executeFetchRequest:toFetchRequest error:&error];
        
        return toEntities;
    }
    @catch (NSException *exception) {
        NSLog(@"Exception while fetching with property %@ on %@: %@", propertyName, entityName, [exception debugDescription]);
    }
    
    return toEntities;
}

-(NSManagedObject*) createEntityFromResultSet:(FMResultSet*) results
                                withTableInfo:(SQCDTableInfo*) tableInfo
                       inManagedObjectContext:(NSManagedObjectContext*) moc
                       shouldCreateManyToMany:(BOOL) createM2M
{
    
    NSString* entityName = [tableInfo representedClassName];
    
    NSString *primaryProperty = [[tableInfo primaryColumn] nameForProperty];
    
    NSArray* existingEntities = [self fetchEntity:entityName
                                         primaryPropertyName:primaryProperty
                                        primaryPropertyValue:[results intForColumn:primaryProperty]
                                      inManagedObjectContext:moc];
    
    if (existingEntities.count > 0) {
        NSLog(@"Found Entity %@", entityName);
        return [existingEntities objectAtIndex:0];
    }
    
    NSLog(@"Creating Entity %@", entityName);
    
    id entity = [NSEntityDescription insertNewObjectForEntityForName:entityName inManagedObjectContext:moc];
    
    for (SQCDColumnInfo* columnInfo in [tableInfo.columns allValues]) {
        
        @autoreleasepool {
            SQCDForeignKeyInfo* foreignKeyInfo = [tableInfo.foreignKeys valueForKey:columnInfo.sqliteName];
            
            if (foreignKeyInfo != nil && ![foreignKeyInfo.toSqliteTableName isEqualToString:tableInfo.sqliteName]) {
                NSLog(@"Ingnoring foreignkey %@ from %@ to %@", foreignKeyInfo.fromSqliteColumnName, foreignKeyInfo.fromSqliteTableName, foreignKeyInfo.toSqliteTableName);
            }else{
                NSString* propertyName = [columnInfo nameForProperty];
                NSString* propertyType = [SQCDTypeMapper xctypeFromType:columnInfo.sqlliteType];
                
                id value = [self valueFromResultSet:results
                                                      forColumnName:columnInfo.sqliteName
                                                       propertyType:propertyType];
                @try {
                    [entity setValue:value forKey:propertyName];
                }
                @catch (NSException *exception) {
                    NSLog(@"Exception while setting %@ on %@: %@", propertyName, entityName, [exception debugDescription]);
                }
            }
        }
    }
    
    if (entity != nil) {
        @autoreleasepool {
            [self createRelationshipForEntity:entity
                                      inManagedObjectContext:moc
                                                   tableInfo:tableInfo
                                                   resultSet:results];
            
        }
        if (createM2M) {
            
            @autoreleasepool {
                [self createManyToManyFromEntity:entity
                                               withTableInfo:tableInfo
                                      inManagedObjectContext:moc];
            }
        }
        
    }
    
    return entity;
}

- (void) createManyToManyFromEntity:(NSManagedObject*) entity
                         withTableInfo:(SQCDTableInfo*) tableInfo
                inManagedObjectContext:(NSManagedObjectContext*) moc
{
    FMDatabase* database = [self openDatabase];

    NSString *primaryProperty = [[tableInfo primaryColumn] nameForProperty];

    NSDictionary *tablesDict = [self tableDictionary];
    
    NSDictionary* inverseDict = [tableInfo.helper inverseRelationships];
    
    NSArray* inverseForTable = [inverseDict objectForKey:tableInfo.sqliteName];
    
    for (SQCDForeignKeyInfo* inverseInfo in inverseForTable) {
        SQCDTableInfo* inverseSourceTableInfo = [tablesDict objectForKey:inverseInfo.toSqliteTableName];
        if ([inverseSourceTableInfo isManyToMany]) {
            // TODO migrate many to many
            SQCDForeignKeyInfo* m2mInfo = [tableInfo.helper manyToManyRelationFromTable:tableInfo.sqliteName
                                                                                  toTable:inverseSourceTableInfo.sqliteName];
            // Fetch results from db and create entities
            int fromColunmValue = [[entity valueForKey:primaryProperty] intValue];
            
            FMResultSet *resultSet = [database executeQuery:[NSString stringWithFormat:@"select * from %@ where %@=%d", inverseSourceTableInfo.sqliteName, inverseInfo.toSqliteColumnName, fromColunmValue]];
            
            NSUInteger count = [database intForQuery:[NSString stringWithFormat:@"select count(%@) from %@ where %@=%d",inverseInfo.toSqliteColumnName,inverseSourceTableInfo.sqliteName, inverseInfo.toSqliteColumnName, fromColunmValue]];
            
            NSMutableArray* relationEntities = [NSMutableArray array];
            
            SQCDTableInfo* toTableInfo = [tablesDict objectForKey:m2mInfo.toSqliteTableName];
            
            @autoreleasepool {
                while([resultSet next]) {
                    
                    int destValue = [resultSet intForColumn:[toTableInfo primaryColumn].sqliteName];
                    FMResultSet *destResults = [database executeQuery:[NSString stringWithFormat:@"select * from %@ where %@=%d ", toTableInfo.sqliteName, [toTableInfo primaryColumn].sqliteName, destValue]];
                    
                    while([destResults next]) {
                        NSManagedObject *relationEntity =  [self createEntityFromResultSet:destResults
                                                                                             withTableInfo:toTableInfo
                                                                                    inManagedObjectContext:moc
                                                                                    shouldCreateManyToMany:NO];
                        [relationEntities addObject:relationEntity];
                    }
                }
            }
            
            @try {
                NSLog(@"Creating Relation %@ form %@ to %@", m2mInfo.relationName, m2mInfo.fromSqliteTableName, m2mInfo.toSqliteTableName);
                // Set the reation property upon the fromEntity
                if (relationEntities.count == count) {
                    [entity setValue:[NSSet setWithArray:relationEntities] forKey:m2mInfo.relationName];
                }
            }
            @catch (NSException *exception) {
                NSLog(@"Exception while setting %@ on %@: %@", m2mInfo.relationName, entity.entity.name, [exception debugDescription]);
            }
        }
    }
    
//    [database close];

}

-(void) createRelationshipForEntity:(NSManagedObject*) fromEntity
             inManagedObjectContext:(NSManagedObjectContext*) moc
                          tableInfo:(SQCDTableInfo*) tableInfo
                          resultSet:(FMResultSet*) fromResultSet
{
    FMDatabase* database = [self openDatabase];
    
    // Fetch all the related entries from the destination table
    NSArray* foreignKeyInfo = [tableInfo.foreignKeys allValues];
    // Iterate over all foreign keys
    [foreignKeyInfo enumerateObjectsUsingBlock:^(SQCDForeignKeyInfo* relationInfo, NSUInteger idx, BOOL *stop) {
        
        if (![relationInfo.toSqliteTableName isEqualToString:tableInfo.sqliteName]) {
            
            // Get the destination table info
            NSDictionary *tablesDict = [self tableDictionary];
            SQCDTableInfo* toTableInfo = [tablesDict valueForKey:relationInfo.toSqliteTableName];
            
            // Fetch results from db and create entities
            int fromColumnValue = [fromResultSet intForColumn:relationInfo.fromSqliteColumnName];
            FMResultSet *results = [database executeQuery:[NSString stringWithFormat:@"select * from %@ where %@=%d", relationInfo.toSqliteTableName, relationInfo.toSqliteColumnName, fromColumnValue]];
            NSMutableArray* relationEntities = [NSMutableArray array];
            
            while([results next]) {
                NSLog(@"Creating Relation %@ form %@ to %@", relationInfo.relationName, fromEntity.entity.name, [relationInfo.toSqliteTableName capitalizedString]);

                NSManagedObject *relationEntity =  [self createEntityFromResultSet:results
                                                                                    withTableInfo:toTableInfo
                                                                           inManagedObjectContext:moc
                                                                            shouldCreateManyToMany:YES];
                [relationEntities addObject:relationEntity];
            }
            
            @try {
                // Set the reation property upon the fromEntity
                if (relationEntities.count) {
                    if (relationInfo.toMany) {
                        [fromEntity setValue:[NSSet setWithArray:relationEntities] forKey:relationInfo.relationName];
                    } else{
                        [fromEntity setValue:[relationEntities objectAtIndex:0] forKey:relationInfo.relationName];
                    }
                }
            }
            @catch (NSException *exception) {
                NSLog(@"Exception while setting %@ on %@: %@", relationInfo.relationName, fromEntity.entity.name, [exception debugDescription]);
            }
        }
    }];
    
//    [database close];
}

#pragma mark - DB helpers

- (id) valueFromResultSet:(FMResultSet*) results
            forColumnName:(NSString*) columnName
             propertyType:(NSString*) propertyType
{
    id value = nil;
    
    // TODO handle other values
    if ([propertyType isEqualToString:XCSTRING]) {
        value = [results stringForColumn:columnName];
    } else if ([propertyType isEqualToString:XCINT64] || [propertyType isEqualToString:XCINT32]){
        value = [NSNumber numberWithInt:[results intForColumn:columnName]];
    } else if ([propertyType isEqualToString:XCDECIMAL]){
        value = [NSDecimalNumber numberWithFloat:[[results stringForColumn:columnName] floatValue]];
    } else if ([propertyType isEqualToString:XCDATE]){
        value = [results dateForColumn:columnName];
    }else{
        value = [results objectForColumnName:columnName];
    }
    
    if ([value isKindOfClass:[NSNull class]]) {
        value = nil;
    }
    
    return value;
}

- (FMDatabase *)openDatabase
{
    if (self.dataBase) {
        return self.dataBase;
    }
    self.dataBase = [FMDatabase databaseWithPath:self.dbPath];
    if (![self.dataBase open]) {
        self.dataBase = nil;
        NSLog(@"Failed to open database!");
    }
    return self.dataBase;
}



@end

File: Sources/SQCDHelper/SQCDMigrationManager.swift
import CoreData
import SqliteExtractor
import _SqliteExtractor_constant
import Foundation
import GRDB

public class SQCDMigrationManager {

    var dbPath = ""
    let dbQueue: DatabaseQueue
    var tableDictionary = [String: SQCDTableInfo]()
    var helper = SQCDDatabaseHelper()


    init(dbPath: String = "", dbQueue: DatabaseQueue, tableDictionary: [String : SQCDTableInfo] = [String: SQCDTableInfo](), helper: SQCDDatabaseHelper = SQCDDatabaseHelper()) {
        self.dbPath = dbPath
        self.dbQueue = dbQueue
        self.tableDictionary = tableDictionary
        self.helper = helper
    }

    public
    static func startDataMigrationWithDBPath(_ dbPath: String!, momdPath momnPath: String!, outputPath: String!, helper: SQCDDatabaseHelper) -> Bool {

        let result = Result {
            var config = GRDB.Configuration()
            config.readonly = true
            config.foreignKeysEnabled = true
            return try DatabaseQueue(path: dbPath, configuration: config)
        }
        guard case let .success(success) = result else {
            NSLog("Failed to open GRDB database at path: \(dbPath)")
            return false
        }
        let manager = SQCDMigrationManager(dbQueue: success)
        manager.helper = helper
        manager.dbPath = dbPath


        do {
            manager.tableDictionary = try manager.helper.fetchTableInfos(dbPath)
        } catch {
            NSLog("Failed to fetch table infos: \(error)")
            return false
        }

        let cdm = SQCDCoreDataManager(withModelPath: momnPath, outputDirectory: outputPath)
        let moc: NSManagedObjectContext! = cdm.managedObjectContext
        do {
            try manager.migrateTableDataFromDatabase(toManagedObjectContext: moc, withRelationships: true)
            return true
        } catch {
            dump(error)
            return false
        }
    }

    func migrateTableDataFromDatabase(toManagedObjectContext moc: NSManagedObjectContext!, withRelationships migrateRelationships: Bool) throws  {
        let tablesDict = self.tableDictionary
        for tableInfo in tablesDict.values {
            let tableName: String! = tableInfo.sqliteName

            try autoreleasepool {
                if tableName != "sqlite_sequence" && !tableInfo.isManyToMany() {
                    NSLog("***************  Started migration for table %@  ****************", tableName)
                    try self.dbQueue.read { db in
                        let cursor = try Row.fetchCursor(db, sql: "SELECT * FROM \(tableName!)")

                        while let row = try cursor.next() {
                            try autoreleasepool {
                                let _ = try self.createEntityFromRow(db, row, withTableInfo: tableInfo, inManagedObjectContext: moc, shouldCreateManyToMany: true)
                                try moc.save()
                            }
                        }
                    }

                    NSLog("***************  Ended migration for table %@  ****************", tableName)
                }
            }
        }
    }

    func createEntityFromRow(_ dataBase:GRDB.Database , _ row: Row, withTableInfo tableInfo: SQCDTableInfo, inManagedObjectContext moc: NSManagedObjectContext!, shouldCreateManyToMany createM2M: Bool) throws -> NSManagedObject? {
        let entityName = tableInfo.representedClassName()
        let primaryProperty = tableInfo.primaryColumn()?.nameForProperty()
        let primaryValue = row[primaryProperty!] as NSNumber? ?? (-1 as NSNumber)

        let request = NSFetchRequest<NSManagedObject>(entityName: entityName)
        let predicateStr = String(format: "(%@ IN %@)", primaryProperty!, "%@")
        request.predicate = NSPredicate(format: predicateStr, [primaryValue])
        let existing = try moc.fetch(request)
        if existing.count > 0 {
            return existing.first
        }

        let entity = NSEntityDescription.insertNewObject(forEntityName: entityName, into: moc)

        for columnInfo in tableInfo.columns.values {
            autoreleasepool {
                let foreignKeyInfo = tableInfo.foreignKeys[columnInfo.sqliteName]
                if let foreignKeyInfo, foreignKeyInfo.toSqliteTableName != tableInfo.sqliteName {
                    return
                }

                let propertyName = columnInfo.nameForProperty()
                let propertyType = SQCDTypeMapper.xctypeFromType(columnInfo.sqlliteType)
                let value = self.valueFromRow(row, columnName: columnInfo.sqliteName, propertyType: propertyType)

                entity.setValue(value, forKey: propertyName)
            }
        }

        try self.createRelationshipForEntity(dataBase,entity, inManagedObjectContext: moc, tableInfo: tableInfo, sourceRow: row)
        if createM2M {
            try self.createManyToManyFromEntity(dataBase,entity, withTableInfo: tableInfo, inManagedObjectContext: moc, sourceRow: row)
        }

        return entity
    }

    func createRelationshipForEntity(_ db: GRDB.Database,_ fromEntity: NSManagedObject!, inManagedObjectContext moc: NSManagedObjectContext!, tableInfo: SQCDTableInfo, sourceRow: Row) throws {
        let foreignKeys = tableInfo.foreignKeys.values
        for relationInfo in foreignKeys where relationInfo.toSqliteTableName != tableInfo.sqliteName {
            guard let toTableInfo = tableDictionary[relationInfo.toSqliteTableName] else { continue }
            let fromColumnValue: CInt = sourceRow[relationInfo.fromSqliteColumnName] ?? -1
            let cursor = try Row.fetchCursor(db, sql: "SELECT * FROM \(relationInfo.toSqliteTableName) WHERE \(relationInfo.toSqliteColumnName) = ?", arguments: [fromColumnValue])

            var relatedEntities: [NSManagedObject] = []

            while let result = try cursor.next() {
                if let related = try createEntityFromRow(db, result, withTableInfo: toTableInfo, inManagedObjectContext: moc, shouldCreateManyToMany: true) {
                    relatedEntities.append(related)
                }
            }

            if !relatedEntities.isEmpty {
                if relationInfo.toMany {
                    fromEntity.setValue(NSSet(array: relatedEntities), forKey: relationInfo.relationName)
                } else {
                    fromEntity.setValue(relatedEntities.first, forKey: relationInfo.relationName)
                }
            }
        }
    }

    func createManyToManyFromEntity(_ db:GRDB.Database , _ entity: NSManagedObject!, withTableInfo tableInfo: SQCDTableInfo, inManagedObjectContext moc: NSManagedObjectContext!, sourceRow: Row) throws {
        let primaryProperty = tableInfo.primaryColumn()?.nameForProperty()
        guard let primaryValue = entity.value(forKey: primaryProperty!) as? CInt else { return }

        let inverseDict = tableInfo.helper.inverseRelationships
        let inverseForTable = inverseDict[tableInfo.sqliteName] ?? []

        for inverseInfo in inverseForTable {
            guard let inverseTableInfo = tableDictionary[inverseInfo.toSqliteTableName],
                  inverseTableInfo.isManyToMany(),
                  let m2mInfo = tableInfo.helper.manyToManyRelationFromTable(tableInfo.sqliteName, toTableName: inverseTableInfo.sqliteName),
                  let toTableInfo = tableDictionary[m2mInfo.toSqliteTableName] else { continue }
            let junctionCursor = try Row.fetchCursor(db, sql: "SELECT * FROM \(inverseTableInfo.sqliteName) WHERE \(inverseInfo.toSqliteColumnName) = ?", arguments: [primaryValue])

            var relatedEntities: [NSManagedObject] = []
            while let junctionRow = try junctionCursor.next() {
                guard let destNumber = junctionRow[toTableInfo.primaryColumn()!.sqliteName] as NSNumber? else {
                    NSLog("Warning: No valid destination ID found in junction table for entity: \(entity), table: \(toTableInfo.sqliteName)")
                    assertionFailure()
                    continue
                }
                let destId: CInt = destNumber.int32Value

                let destCursor = try Row.fetchCursor(db, sql: "SELECT * FROM \(toTableInfo.sqliteName) WHERE \(toTableInfo.primaryColumn()!.sqliteName) = ?", arguments: [destId])
                while let destRow = try destCursor.next() {
                    if let related = try createEntityFromRow(db, destRow, withTableInfo: toTableInfo, inManagedObjectContext: moc, shouldCreateManyToMany: false) {
                        relatedEntities.append(related)
                    }
                }
            }

            if relatedEntities.count > 0 {
                entity.setValue(NSSet(array: relatedEntities), forKey: m2mInfo.relationName)
            }
        }
    }

    func valueFromRow(_ row: Row, columnName: String, propertyType: String) -> NSObjectProtocol? {
        var value: NSObjectProtocol?

        switch propertyType {
        case XCSTRING:
            value = row[columnName] as NSString?
        case XCINT32, XCINT64, XCFLOAT , XCINT16, XCDOUBLE:
            if let numbering:NSNumber? = row[columnName]  {
                value = numbering
            }
        case XCDECIMAL:
            if let str: Decimal = row[columnName] {
                value = str as NSDecimalNumber
            }
        case XCDATE:
            value = row[columnName] as NSDate?
        case XCBINARY:
            if let data:Data = row[columnName] {
                value = data as NSData
            }
        case XCBOOL:
            if let boolValue: Bool = row[columnName] {
                value = NSNumber(value: boolValue)
            }
            
        default:
            value = row[columnName] as? NSObjectProtocol
        }

        return value is NSNull ? nil : value
    }
}
